File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        
        lm,rm = [0]*(len(height)),[0]*(len(height))
        ans = 0
            
        for i in range(len(rm)-2,-1,-1):
            rm[i] = max(rm[i+1],height[i+1])
            
        for i in range(1,len(lm)):
            lm[i] = max(lm[i-1],height[i-1])
            water = min(lm[i],rm[i]) - height[i]
            ans += water if water>0 else 0
    
        return ans

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int ans=0, leftM = height[0], n = height.size();
        vector<int> rightM(n,height[n-1]);
        for(int i=n-2;i>=0;i--){
            rightM[i] = max(rightM[i+1],(i==n-2?0:height[i+1]));
        }
        
        for(int i=1;i<n;i++){
            leftM = max((i==1?0:height[i-1]),leftM);
            // cout<<leftM<<","<<rightM[i]<<":"<<min(leftM,rightM[i])-height[i]<<endl;
            ans += min(leftM,rightM[i])<height[i]?0:min(leftM,rightM[i])-height[i];
        }
        
        return ans;
            
    }
};